### PR TITLE
Use struct tags

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

File: src/ansi-c/c_typecheck_type.cpp [hunk 1]
```diff
@@ -822,18 +822,17 @@ void c_typecheck_baset::typecheck_compound_type(struct_union_typet &type)
...
-  if(type.id() == ID_union)
-  {
-    union_tag_typet tag_type(identifier);
-    tag_type.add_source_location() = type.source_location();
-    type.swap(tag_type);
-  }
+  typet tag_type;
+
+  if(type.id() == ID_union || type.id()==ID_incomplete_union)
+    tag_type=union_tag_typet(identifier);
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_type.cpp]: Nit pick: spacing around `=` (here and below).

File: src/ansi-c/c_typecheck_code.cpp [hunk 2]
```diff
@@ -383,7 +383,9 @@ bool c_typecheck_baset::is_complete_type(const typet &type) const
   }
   else if(type.id()==ID_vector)
     return is_complete_type(type.subtype());
-  else if(type.id() == ID_symbol_type)
+  else if(
+    type.id() == ID_symbol_type || type.id() == ID_struct_tag ||
+    type.id() == ID_union_tag)
```
Comment (tautschnig) [inline src/ansi-c/c_typecheck_code.cpp:388]: Please add braces around the (one-line) body for readability.

**Review by tautschnig — APPROVED** [review]
Approving with some formatting requests (and assuming CI will pass).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a21cd682877b7088185e8282839940ad05c8ebfd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85338055

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 04c26c4f688131a695b266817ed3bc81d956d781).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85349549

**Comment by kroening** [comment]
It should be mentioned that the transition to struct and union tags was a ~5 year project!

**Review by smowton — APPROVED** [review]

**Review by chrisr-diffblue — APPROVED** [review]

**Pull request merged**